cmd/snap-confine/udev-support.c

```diff
@@ -18,317 +18,436 @@
-#include <sys/wait.h>
 #include <unistd.h>
 
+#include <libudev.h>
+
+#include "../libsnap-confine-private/cleanup-funcs.h"
 #include "../libsnap-confine-private/snap.h"
 #include "../libsnap-confine-private/string-utils.h"
+#include "../libsnap-confine-private/cgroup-support.h"
 #include "../libsnap-confine-private/utils.h"
 #include "udev-support.h"
 
-static void
-_run_snappy_app_dev_add_majmin(struct snappy_udev *udev_s,
-			       const char *path, unsigned major, unsigned minor)
+__attribute__((format(printf, 2, 3)))
+static void sc_dprintf(int fd, const char *format, ...);
+
+static void sc_dprintf(int fd, const char *format, ...)
 {
-	int status = 0;
-	pid_t pid = fork();
-	if (pid < 0) {
-		die("cannot fork support process for device cgroup assignment");
+	va_list ap1;
+	va_list ap2;
+	int n_expected, n_actual;
+
+	va_start(ap1, format);
+	va_copy(ap2, ap1);
+	n_expected = vsnprintf(NULL, 0, format, ap2);
+	n_actual = vdprintf(fd, format, ap1);
+	if (n_actual == -1 || n_expected != n_actual) {
+		die("cannot write to fd %d", fd);
 	}
-	if (pid == 0) {
-		uid_t real_uid, effective_uid, saved_uid;
-		if (getresuid(&real_uid, &effective_uid, &saved_uid) != 0)
-			die("cannot get real, effective and saved user IDs");
-		// can't update the cgroup unless the real_uid is 0, euid as
-		// 0 is not enough
-		if (real_uid != 0 && effective_uid == 0)
-			if (setuid(0) != 0)
-				die("cannot set user ID to zero");
-		char buf[64] = { 0 };
-		// pass snappy-add-dev an empty environment so the
-		// user-controlled environment can't be used to subvert
-		// snappy-add-dev
-		char *env[] = { NULL };
-		if (minor == UINT_MAX) {
-			sc_must_snprintf(buf, sizeof(buf), "%u:*", major);
-		} else {
-			sc_must_snprintf(buf, sizeof(buf), "%u:%u", major,
-					 minor);
-		}
-		debug("running snap-device-helper add %s %s %s",
-		      udev_s->tagname, path, buf);
-		// This code runs inside the core snap. We have two paths
-		// for the udev helper.
-		//
-		// First try new "snap-device-helper" path first but
-		// when running against an older core snap fallback to
-		// the old name.
-		if (access("/usr/lib/snapd/snap-device-helper", X_OK) == 0)
-			execle("/usr/lib/snapd/snap-device-helper",
-			       "/usr/lib/snapd/snap-device-helper", "add",
-			       udev_s->tagname, path, buf, NULL, env);
-		else if (access("/usr/libexec/snapd/snap-device-helper", X_OK) == 0)
-			execle("/usr/libexec/snapd/snap-device-helper",
-			       "/usr/libexec/snapd/snap-device-helper", "add",
-			       udev_s->tagname, path, buf, NULL, env);
-		else
-			execle("/lib/udev/snappy-app-dev",
-			       "/lib/udev/snappy-app-dev", "add",
-			       udev_s->tagname, path, buf, NULL, env);
-		die("execl failed");
+	va_end(ap2);
+	va_end(ap1);
+}
+
+/* Allow access to common devices. */
+static void sc_udev_allow_common(int devices_allow_fd)
+{
+	/* The devices we add here have static number allocation.
+	 * https://www.kernel.org/doc/html/v4.11/admin-guide/devices.html */
```

Comment:
just a note that the other comments reference https://www.kernel.org/doc/Documentation/admin-guide/devices.txt while this one is a different link, would be kinda nice if they all referenced the same link, I have no strong preference for which link we choose just that we are consistent

again for a followup